gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    public Builder<ElementT, ElementResultT, RequestT, ResponseT> setPrototype(RequestT prototype) {
+      this.prototype = prototype;
+      return this;
+    }
+
+    public BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT> build() {
+      return new BatcherImpl<>(this);
+    }
+  }
+
+  public static <EntryT, EntryResultT, RequestT, ResponseT>
+      Builder<EntryT, EntryResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher.");
+
+    if (currentOpenBatch == null) {
+      currentOpenBatch = new Batch<>(batchingDescriptor.newRequestBuilder(prototype));
+    }
+
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    currentOpenBatch.add(element, result);
+    return result;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() throws InterruptedException {
+    sendBatch();
+    awaitAllOutstandingBatches();
+  }
+
+  /** Sends accumulated elements asynchronously for batching. */
+  private void sendBatch() {
+    if (currentOpenBatch == null) {
+      return;
+    }
+    final Batch<ElementT, ElementResultT, RequestT> accumulatedBatch = currentOpenBatch;
+    currentOpenBatch = null;
+    numOfOutstandingBatches.incrementAndGet();
+
+    final ApiFuture<ResponseT> batchResponse =
+        callable.futureCall(accumulatedBatch.builder.build());
+
+    ApiFutures.addCallback(
+        batchResponse,
+        new ApiFutureCallback<ResponseT>() {
+          @Override
+          public void onSuccess(ResponseT response) {
+            try {
+              batchingDescriptor.splitResponse(response, accumulatedBatch.results);
+            } catch (Throwable ex) {
+              for (SettableApiFuture<ElementResultT> result : accumulatedBatch.results) {
+                result.setException(ex);
+              }
+            } finally {
+              onBatchCompletion();
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable throwable) {
+            try {
+              batchingDescriptor.splitException(throwable, accumulatedBatch.results);
+            } catch (Throwable ex) {
+              for (SettableApiFuture<ElementResultT> result : accumulatedBatch.results) {
+                result.setException(ex);
+              }
+            } finally {
+              onBatchCompletion();
+            }
+          }
+        },
+        directExecutor());
+  }
+
+  private void onBatchCompletion() {
+    if (numOfOutstandingBatches.decrementAndGet() == 0) {
+      synchronized (flushLock) {
+        flushLock.notifyAll();
+      }
+    }
+  }
+
+  private void awaitAllOutstandingBatches() throws InterruptedException {
+    while (numOfOutstandingBatches.get() > 0) {
+      synchronized (flushLock) {
+        flushLock.wait(DEFAULT_WAIT_TIME_MS);
```

Comment:
I'm not sure I understand why we need a timeout

Reply:
I added this as a defense, Also inspired by existing [CBT-OperationalAccountant](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/OperationAccountant.java#L97-L102)

Reply:
What is it defending against?

Reply:
I added this for the case of RPC taking too long, which would enable this thread to keep on checking the numOfOutStandingBatch.

I now realized that this would not add any meaning. The only update we are waiting here would be from onBatchComplete's notifyAll(), which will execute on all cases.
Thanks a lot for pointing it out. I will remove this timeout.